feat: 질문 API 구현

c-yeonwoo opened the pull request:
## Resolve #59 

- 면접질문(Question) 도메인에 이어서 화면으로부터 요청을 받아들여 로직을 수행할 API단 구현이 필요하다.

## Changes

- QuestionController, QuestionControllerTest 작성
- QuestionService, QuestionServiceTest 작성
- CRUD 및 검색기능 구현
- 테스트에서 한글 인코딩을 위한 설정 추가

## Notes
- 인수테스트 추가 작성하여 푸시할 예정

## References

- N/A

c-yeonwoo committed:
feat: QuestionController 구현 및 테스트

- QuestionController 생성
- QuestionControllerTest 생성
- CRUD 구현 및 테스트

c-yeonwoo committed:
feat: QuestionService 구현 및 테스트

- QuestionService 생성
- QuestionServiceTest 생성
- CRUD 구현 및 테스트

c-yeonwoo committed:
feat: 질문 검색 기능 구현

- controller에서 readAll, searchByTitle로 메서드 분기 (같은 url)
- JPQL 쿼리 생성 ("select q from Question q where title like %:keyword%")
- 기능 추가에 따른 테스트 코드 작성

c-yeonwoo committed:
feat: 테스트 환경에서 한글을 지원하기 위해 mockMvc 설정 추가

- MvcTest에 인코딩 설정 추가

c-yeonwoo committed:
feat: AcceptanceTest에 User 인증을 위한 메서드 추가 및 QuestionAcceptanceTest 시나리오 작성

- post, put, patch, delete
- bearerToken 부분 추가

c-yeonwoo committed:
feat: 인수테스트를 위한 token 생성

- test > resources > data.sql에 user 데이터 생성

c-yeonwoo committed:
feat: interceptor를 통과하지 않도록 annotaion 기능 추가

hongbin-dev commented on src/main/java/underdogs/devbie/question/controller/QuestionController.java:
```diff
@@ -0,0 +1,92 @@
+        @LoginUser User user,
+        @Valid @RequestBody QuestionCreateRequest request) {
+        Long questionId = questionService.save(user.getId(), request);
+        return ResponseEntity
+            .created(URI.create(String.format("/api/questions/%d", questionId)))
+            .build();
```
String.format() 👍

hongbin-dev commented on src/main/java/underdogs/devbie/question/controller/QuestionController.java:
```diff
@@ -0,0 +1,92 @@
+    @GetMapping
+    public ResponseEntity<QuestionResponses> readAll() {
+        QuestionResponses responses = questionService.readAll();
+        return ResponseEntity
+            .ok()
+            .body(responses);
```
```suggestion
        return ResponseEntity.ok(responses)
```
별건 아니지만 어떻게할지 다같이 정해볼까요?

c-yeonwoo replied:
앨런네는 ok(responses)로 하신건가요?
body에 내려보낸다는 의미가 명확해서 일부러 body로 사용하긴했는데, 
ok안 에 넣는것도 코드가 심플하고 괜찮은것 같네요!

팀에서 합의가 이뤄지면 수정하도록 하겠습니다.

hongbin-dev commented on src/main/java/underdogs/devbie/question/controller/QuestionController.java:
```diff
@@ -0,0 +1,92 @@
+            .body(responses);
+    }
+
+    @GetMapping(params = "keyword")
+    public ResponseEntity<QuestionResponses> searchByTitle(
+        @RequestParam("keyword") String keyword
```
GetMapping에 params라는 필드가 있군요!
하지만 `@RequestParam("keyword") String keyword`이 코드로 생략이 가능하다고 생각되네요.

c-yeonwoo replied:
현재 GET : "/api/questions" 해당 url로 두개의 함수가 매핑되어있습니다.
처음엔 생략하고 시도했더니 spring에서 context load가 실패하더군요
같은 url로 메서드를 오버로드하려면 현재와같은 방법이 최선일 것 같은데 더 나은방법이 있을까요?

hongbin-dev replied:
아,, 앞에 코드를 보지못하고 리뷰를 남겼네요.
같은 URL에 파라미터로 분기하려면 params를 사용하면 되겠네요. 배워갑니다. 👍👍

hongbin-dev commented on src/main/java/underdogs/devbie/question/controller/QuestionController.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    @PatchMapping("/{id}")
+    public ResponseEntity<Void> update(
+        @LoginUser User user,
+        @PathVariable("id") Long id,
+        @Valid @RequestBody QuestionUpdateRequest request
+    ) {
```
Valid 👍👍

hongbin-dev commented on src/main/java/underdogs/devbie/question/domain/QuestionRepository.java:
```diff
@@ -0,0 +1,12 @@
+import org.springframework.data.repository.query.Param;
+
+public interface QuestionRepository extends JpaRepository<Question, Long> {
+    @Query("select q from Question q where q.title.title like %:keyword%")
+    List<Question> findByTitleLike(@Param("keyword") String keyword);
```
이 부분은 JPQL 편의메서드에서 지원하는것 같은데요.
객체로 한번 더 포장되어서 어려울까요?

[Stackoverflow](https://stackoverflow.com/questions/21456494/spring-jpa-query-with-like)

c-yeonwoo replied:
이 부분도 처음에는 spring data jpa의 Name function을 사용해서
직접 쿼리를 작성하지 않아도 findByTitleLike~로 가능할 줄 알았는데요
예상하신 것과 같이 객체로 매핑되어있어서 결과가 올바르게 나오지 않는것같습니다.
인자로 String이 아닌 wrapper class를 넣으라고 하더군요 이름은 title인데도 말이죠

hongbin-dev commented on src/main/java/underdogs/devbie/question/dto/QuestionResponses.java:
```diff
@@ -0,0 +1,21 @@
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@AllArgsConstructor(access = AccessLevel.PROTECTED)
+@Getter
+public class QuestionResponses {
```
### 
```suggestion
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@Getter
public class QuestionResponses {
```
정적팩토리 메서드가 있는데 PROTECTED로 하신 이유가 있을까요?

c-yeonwoo replied:
큰 생각없이 복붙하다보니 이런 실수가 나왔네요 정정하겠습니다!

hongbin-dev commented on src/test/java/underdogs/devbie/MvcTest.java:
```diff
@@ -1,19 +1,35 @@
+    @Autowired
+    private WebApplicationContext ctx;
+
+    @BeforeEach
+    void setUpEncoding() {
+        mockMvc = MockMvcBuilders.webAppContextSetup(ctx)
+            .addFilters(new CharacterEncodingFilter("UTF-8", true))
+            .alwaysDo(print())
+            .build();
+    }
```
👍👍👍

hongbin-dev commented on src/test/java/underdogs/devbie/acceptance/AcceptanceTest.java:
```diff
@@ -40,6 +42,21 @@ private static RequestSpecification given() {
 
+    protected <T> void post(String path, String inputJson, String bearerToken) {
+        // @formatter:off
+        given().
+                header(AUTH_HEADER, bearerToken).
+                body(inputJson).
+                contentType(MediaType.APPLICATION_JSON_VALUE).
+                accept(MediaType.APPLICATION_JSON_VALUE).
```
```suggestion
        given().
                .auth().oauth2(bearerToken)
                body(inputJson).
                contentType(MediaType.APPLICATION_JSON_VALUE).
                accept(MediaType.APPLICATION_JSON_VALUE).
```
이건 어떨까요~?

c-yeonwoo replied:
저렇게 하면 header의 "Authorization" 키밸류로 자동 매핑이되나요?🤓

hongbin-dev replied:
네!! bearer도 뺄 수 있어서 좋아요. 토큰 값만 넣어도되서 심플하더라구요!

hongbin-dev commented on src/test/java/underdogs/devbie/question/acceptance/QuestionAcceptanceTest.java:
```diff
@@ -0,0 +1,38 @@
+
+    private final ObjectMapper objectMapper = new ObjectMapper();
+
+    /*
+    Feature: 질문 관리
+
+    Scenario: 질문을 관리한다.
+            WHEN 회원 가입을 한다.
+            THEN 유저를 반환한다.
+
+            WHEN 질문 생성 요청을 한다.
+            THEN 질문이 생성 된다.
+
+            WHEN 질문 목록 조회를 요청한다.
+            THEN 생성된 질문 목록을 불러온다.
+
+            WHEN 질문 상세 조회를 요청을 한다.
+            THEN 질문 상세 정보를 확인한다.
+
+            WHEN 질문의 제목과 내용을 수정을 요청한다.
+            THEN 질문 제목과 내용이 수정된다.
+
+            WHEN 질문의 제목에 포함된 단어로 검색한다.
+            THEN 해당 단어가 포함된 질문 목록이 조회된다.
+
+            WHEN 질문을 삭제 요청한다..
+            THEN 질문이 삭제된다.
+     */
```
👍👍👍

hongbin-dev commented on src/test/java/underdogs/devbie/question/controller/QuestionControllerTest.java:
```diff
@@ -0,0 +1,156 @@
+    @MockBean
+    private QuestionService questionService;
+
+    @MockBean
+    private BearerAuthInterceptor bearerAuthInterceptor;
+
+    @MockBean
+    private LoginUserArgumentResolver loginUserArgumentResolver;
```
```suggestion
    @Autowired
    private BearerAuthInterceptor bearerAuthInterceptor;

    @Autowired
    private LoginUserArgumentResolver loginUserArgumentResolver;
```
로그인 했다고 가정되는데요! 찐 Bean을 써야하지 않나 싶어요.

c-yeonwoo replied:
setUp부분에 보면 Interceptor와 resolver에서 mock으로 처리해야할 부분이 존재합니다.
그렇지 않으면, 여기서도 인수테스트 처럼 진짜 토큰을 가지고 테스트해야 합니다.

hongbin-dev reviewed (commented):
안녕하세요~ 보스독. 열심히 코드 작성해주셨군요!
가벼운 코멘트 남깁니다~!  꼼꼼한 코드에 감탄하고 가요!
감사합니다~

c-yeonwoo committed:
refactor: UserResponse 생성자 접근제한자 private으로 변경

c-yeonwoo committed:
refactor: BearerIntercepter

- class에도 어노테이션 붙일수 있도록 타입 추가
- 부정형 표현 수정

c-yeonwoo committed:
refactor: Dto 생성자 접근제한자 수정

- AllArgsConstructor : Protected -> Private

c-yeonwoo committed:
Merge branch 'feature/70-acceptance_user' of https://github.com/woowacourse-teams/2020-devbie into feature/59-question_api

c-yeonwoo committed:
test: QuestionAcceptanceTest 테스트 코드 추가

- CRUD + Search
- AcceptanceTest에 토큰 받는 부분 수정

c-yeonwoo committed:
refactor: traivs datasource 추가

c-yeonwoo committed:
refactor: travis datasource 제거

c-yeonwoo committed:
feat: data.sql 제거 및 User 생성 api 추가

- AcceptanceTest에 createUser 메서드 추가
- User를 먼저 만들어 놓고, 생성한 userId로 토큰을 만든다.

c-yeonwoo commented:
죄송합니다. 커밋이 좀 지저분하네요 ㅠ

hongbin-dev reviewed (approved):
수고하셨습니다.

c-yeonwoo committed:
feat: AcceptanceTest 정리

- build.gradle에 travis log 찍는 코드 추가
- user 생성 및 토큰 전역으로 사용
- 토큰생성을 위한 유효기간 늘림(300 -> 30000)

slowCoyle commented on build.gradle:
```diff
@@ -75,6 +76,10 @@ dependencies {
 
 test {
     useJUnitPlatform()
+    testLogging {
+        events "passed", "skipped", "failed"
+        exceptionFormat "full"
+    }
```
이건 무언가요 ?

c-yeonwoo replied:
travis쪽에서 테스트 통과 여부를 로깅하는 코드입니다.
테스트가 실패했을때, 어떤 테스트가 어떻게 실패했는지 로깅을 볼 수 있어서 나름 유용한것 같습니다!

slowCoyle commented on src/main/java/underdogs/devbie/auth/controller/interceptor/BearerAuthInterceptor.java:
```diff
@@ -19,6 +20,11 @@
     public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) {
+        InterceptorValidator interceptorValidator = new InterceptorValidator();
+
+        if (interceptorValidator.isNotValid(handler)) {
+            return true;
+        }
```
제 브랜치에서도 위처럼 변경해주어야 충돌이 안나겠죠 ?

c-yeonwoo replied:
옙 아마 그럴거에요. 
지금 저부분은 pr이 머지되었으니,develop pull한번 받으셔도 좋을것 같네요

slowCoyle commented on src/main/java/underdogs/devbie/question/exception/NotMatchedAuthorException.java:
```diff
@@ -0,0 +1,7 @@
+package underdogs.devbie.question.exception;
+
+public class NotMatchedAuthorException extends RuntimeException {
```
`NotMatchedQuestionAuthorException`으로 이름을 변경하면 어떨까요 ?

Answer에서 사용하는 Exception 중 동일한 Exception이 존재해서 혼란스러울 수도 있을 것 같네요

전 `NotMatchedAnswerAuthorException`로 명명하였습니다.

c-yeonwoo replied:
옙 반영하겠습니다

slowCoyle commented on src/main/java/underdogs/devbie/question/service/QuestionService.java:
```diff
@@ -0,0 +1,78 @@
+    public void update(Long userId, Long questionId, QuestionUpdateRequest request) {
+        Question question = readOne(questionId);
+
+        if (isNotAuthorOfQuestion(userId, question)) {
+            throw new NotMatchedAuthorException();
+        }
```
아래에도 쓰이는 로직인데 함수화를 하는 것은 어떨까요 ?

c-yeonwoo replied:
중복을 못보았군요! 좋은 지적 감사드립니다

c-yeonwoo committed:
refactor: exception naming 수정, 중복되는 코드 함수추출

c-yeonwoo committed:
refactor: ResponseEntity.ok() 리턴 시 컨벤션 수정

- ResponseEntity.ok().body(response) -> RequestEntity.ok(response);

c-yeonwoo committed:
refactor: travis logging 설정 변경

- 실패할 때만 log 출력

sonypark commented on src/test/java/underdogs/devbie/acceptance/AcceptanceTest.java:
```diff
@@ -9,40 +9,55 @@
-        UserTokenDto userTokenDto = UserTokenDto.from(User.builder().id(1L).build());
+
+        Long userId = createUser();
+        UserTokenDto userTokenDto = UserTokenDto.from(User.builder().id(userId).build());
```
여기 빌더 패턴도 아래와 같이 포맷을 맞추면 좋겠네요 :)
```suggestion
        UserTokenDto userTokenDto = UserTokenDto.from(User.builder()
               .id(userId)
               .build());
```

c-yeonwoo replied:
매의 눈이시군요 👍

sonypark commented on src/main/java/underdogs/devbie/user/controller/UserController.java:
```diff
@@ -1,22 +1,43 @@
+        return ResponseEntity
+            .ok(UserResponse.from(user));
+    }
+
+    @NoValidate
+    @PostMapping
+    public ResponseEntity<Long> saveUser(@Valid @RequestBody UserCreateRequest request) {
+        Long userId = userService.saveWithoutOAuthId(request);
+        return ResponseEntity
+            .created(URI.create(String.format("/api/users/%d", userId)))
+            .body(userId);
```
이 부분에서 @NoValidate이 필요한건가요?

c-yeonwoo replied:
acceptance 테스트에서 user 만들려면, interceptor를 안타는 api가 필요합니다. (data.sql에서 insert 불가하기 때문)
현재는 임시방편이고, 
제대로 서비스 할때는 아마  admin 전용 controller를 만들던지, 저 api는 외부에서 접근 못하게 노출하지 않는다던지하는 방법이 필요할 것으로 보입니다.

아마 인가 로직 구현할 때 리팩토링 되지않을까 싶네요 :)

sonypark reviewed (approved):
보스독 구현 잘 하셨네요!
@NoValidate 관련해서 코멘트 하나 남겼으니 확인부탁드립니다 :)

c-yeonwoo committed:
refactor: builder 패턴 줄바꿈 교정

c-yeonwoo committed:
refactor: controller 테스트 assert 방식 수정

- json path 사용

c-yeonwoo committed:
refactor: dto annotation 수정 및 controller test assert 형식 변경

- constructor(no -> required -> all)
- builder
- getter
- setter
- toString
- EqualsAndHashCode

c-yeonwoo committed:
fix: 정적팩토리메서드 변경 부분 테스트 코드에 적용
